spec/controllers/reports_controller_spec.rb

```diff
@@ -0,0 +1,216 @@
+require "rails_helper"
+
+RSpec.describe ReportsController, type: :controller do
+  let!(:current_user){FactoryBot.create :user}
```

Comment:
let khác gì vs let! e?
sao k dùng let e?

Reply:
khi khai báo let thì nó chỉ có giá trị trong mỗi block đó.
Khi khai báo let! thì nó có giá trị trong mọi block it của nó. Bản chất của nó là một khối before ngầm nên nó sẽ được tính toán lại trước mỗi block it.
Em dùng let! tại vì current_user nó còn được sử dụng lại ở trong các block it khác ở dưới ạ

Reply:
giống nhau thế nhỉ
![image](https://user-images.githubusercontent.com/65896888/92223311-7c092500-eeca-11ea-929f-20762907df04.png)

Reply:
chắc ý tưởng trùng nhau a ơi :)))

Reply:
chưa rõ em ơi
lấy cho c ví dụ nhé sự khác nhau giữa lét và let!

Reply:
- Khi khai báo let thì nó chỉ có giá trị trong từng khối it nhất định. Nếu như gọi lại biến let nhiều lần trong nhiều khối it thì nó cũng chỉ khai báo một lần trong khối it đó - lần đầu tiên được gọi. Tức là giá trị của nó sẽ không thay đổi trong block it đó kể từ lần đầu được gọi.
![image](https://user-images.githubusercontent.com/68220979/92340327-185a4400-f0e4-11ea-8628-fec7e9ad91e2.png)
Trong khối it đầu tiên, mặc dù đã được delay 3s nhưng giá trị của current_time vẫn không đổi. Sang đến khối it thứ 2 thì current_time được gọi lại và tính tóan lại nên giá trị của nó được thay đổi.

- Còn let! thì nó được gọi trong một khối before ngầm, do đó kết quả đã được tính toán và cache lại trước khi khối it được thực thi nên giá trị của nó không thay đổi trong các khối it khác nhau.
![image](https://user-images.githubusercontent.com/68220979/92340524-e9909d80-f0e4-11ea-8950-3efeaa9a5640.png)
Như ví dụ ở trên, mặc dù đã được delay 3s trong khối it thứ 2 nhưng giá trị của current_time vẫn không đổi.